### ISSUE TITLE
StrEnum annotation causes wrong unreachability diagnostic

### ISSUE THREAD

**Issue by Molkree** [microsoft/pyright#8658]
## Environment data

-   Pylance version: v2024.7.1 (and v2024.7.104 (pre-release))
-   OS and version: Win11 23H2
-   Python version (& distribution if applicable, e.g. Anaconda): 3.12.4

## Code Snippet

```python
from enum import StrEnum


class MyEnum(StrEnum):
    FOO = "foo"
    BAR = "bar"


action_status: MyEnum = MyEnum.BAR
if action_status:
    print("Reached unreachable code")
```

## Expected behavior

The last line is not marked as unreachable

## Actual behavior

The last line is marked as unreachable:

![image](https://github.com/user-attachments/assets/eac5afd8-11b1-4fc4-96e5-6ec254a42b51)

## Additional info

There's no false warning if I do NOT annotate my variable:
```py
action_status = MyEnum.BAR
```

Even though the inferred type seems to be the same in the tooltip on the next line, it's `(variable) action_status: Literal[MyEnum.BAR]` in both cases.
Additionally, there's no false warning if it's a simple `Enum`. `StrEnum`, `IntEnum`, `Flag` and `IntFlag` all seem to trigger this false warning.

**Comment by Molkree** [comment]
> ```
> if action_status:
>     print("Reached unreachable code")
> ```
> 
> what you meant by the code above? did you mean to write something like ?
> 
> ```
> if action_status == MyEnum.BAR:
>     print("Reached unreachable code")
> ```

Sorry, I guess I reduced the example too much. Original code does something like the following, we have action_status as None initially and then it can become truthy value, hence the simple `if action_status` check:
```py
from enum import StrEnum


class MyEnum(StrEnum):
    FOO = "foo"
    BAR = "bar"


action_status: MyEnum | None = None
try:
    raise Exception
except Exception:
    action_status = MyEnum.FOO
if action_status:
    print("Reached unreachable code")
```

But I believe that the original code snippet shows the problem as well.

**Comment by erictraut** [comment]
Thanks for the bug report. It appears that the `ReprEnum` (from which `IntEnum` and `StrEnum` derive) applies additional "magic" that forwards the `__bool__` call on an enum member to the member's value. I've added special-case logic to handle this case. This will be included in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.375.
